pkg/nfd-master/nfd-master.go

```diff
@@ -432,10 +446,116 @@ func (m *nfdMaster) SetLabels(c context.Context, r *pb.SetLabelsRequest) (*pb.Se
+			if err != nil {
+				return labelReply, err
+			}
+			// Call setTaints() with empty list of taints even though the feature
+			// flag is disabled. This ensures that we delete NFD owned stale taints
+			// when flag got turned off.
+		case false:
+			labelReply, err := m.setTaints(c, emptyTaint, r.NodeName)
+			if err != nil {
+				return labelReply, err
+			}
+		}
 	}
 	return &pb.SetLabelsReply{}, nil
 }
 
+// setTaints sets note taints and annotations based on the taints passed via
+// nodeFeatureRule custom resorce. If empty list of taints is passed, currently
+// NFD owned taints and annotations are removed from the node.
+func (m *nfdMaster) setTaints(c context.Context, taints []corev1.Taint, nodeName string) (*pb.SetLabelsReply, error) {
+	var nodeTaints, annotationTaints []corev1.Taint
+	var val string
+	taintStrs := []string{}
+
+	err := authorizeClient(c, m.args.VerifyNodeName, nodeName)
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	cli, err := m.apihelper.GetClient()
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	// fetch the node object
+	node, err := m.apihelper.GetNode(cli, nodeName)
+	if err != nil {
+		return &pb.SetLabelsReply{}, err
+	}
+
+	// update the taint annotation
+	annotation := node.GetAnnotations()
+	oldAnnotation := annotation[nfdv1alpha1.NodeTaintsAnnotation]
+	switch {
+	case len(taints) > 0:
+		for _, taint := range taints {
+			taintStrs = append(taintStrs, taint.ToString())
+			val = strings.Join(taintStrs, ",")
+		}
+		annotation[nfdv1alpha1.NodeTaintsAnnotation] = val
+	case len(taints) == 0:
+		delete(annotation, nfdv1alpha1.NodeTaintsAnnotation)
+	}
+	node.SetAnnotations(annotation)
+
+	// delete stail taints
+	oldTaints := node.Spec.Taints
+	if val, ok := annotation[nfdv1alpha1.NodeTaintsAnnotation]; ok {
+		sts := strings.Split(val, ",")
+		corev1Taints, _, err := taintutils.ParseTaints(sts)
+		if err != nil {
+			return &pb.SetLabelsReply{}, err
+		}
+		for _, ct := range corev1Taints {
+			annotationTaints = append(annotationTaints,
+				corev1.Taint{
+					Key:    ct.Key,
+					Value:  ct.Value,
+					Effect: ct.Effect})
+		}
+	}
+	for _, nodeTaint := range node.Spec.Taints {
+		// If the list of the passed taints is empty then we delete all
+		// the NFD owned node taints, otherwise we filter out node taints
+		// that don't exist on the NodeFeatureRule CR anymore and try to
+		// delete them.
+		if !taintutils.TaintExists(annotationTaints, &nodeTaint) && len(taints) == 0 || !taintutils.TaintExists(annotationTaints, &nodeTaint) && !taintutils.TaintExists(taints, &nodeTaint) {
+			taintsNew, _ := taintutils.DeleteTaint(node.Spec.Taints, &nodeTaint)
+			node.Spec.Taints = taintsNew
+		}
+	}
+
+	// update the node taints
+	if len(taints) > 0 {
+		for _, taint := range taints {
+			if !taintutils.TaintExists(node.Spec.Taints, &taint) {
+				nodeTaints = append(nodeTaints, taint)
+			}
+			continue
+		}
+	}
+	node.Spec.Taints = append(node.Spec.Taints, nodeTaints...)
```

Comment:
This whole logic looks odd and partly broken. E.g. serializing the new taints into a string and the de-serializing back to see what we want to add looks a bit strange 🧐

What we should do imo is smth like:
1. Get the diff between old and new **nfd managed** taints, basically what was *deleted*
2. Go through the list of *deleted* and remove those from the node
3. Go through the list of all *new* taints and add them to the node if they don't exist

Reply:
> 3\. Go through the list of all _new_ tain

The reason for converting new taints into string is that we want to update taint annotation based on the latest taints passed and annotation is in string.  And we need to first ensure that annotation is up to date and in sync with the CR taints because when deleting a taint from the node we rely on the annotation to know if we are allowed to touch that taint or not. I think the complexity comes from the fact that we can not just compare node taints and CR taints and remove based on the diff of two. There will be taints not owned by NFD as well which makes deletion a bit complicated.
This is how it looks like right now:
```
1. Update the annotation so that we can really trust that it is up to date source of truth
   1.1 convert CR taints in to string
   1.2 compare CR taints string vs node taint annotation to get the diff
   1.3 add/remove diff in annotation

2. Delete stale taints from the node
   2.1 convert updated annotation in to taints array
   2.2 compare taints array vs node taints to get the diff
   2.3 add/remove diff in node taint
```
We can't get diff (2.2) without 1.3. Since annotation is source of truth, the logic first tries to update it and the next steps rely on that updated annotation.

Reply:
> And we need to first ensure that annotation is up to date and in sync with the CR taints

Really? Couldn't you just loop over the old taints and use `taintutils.TaintExists` to see if each of those exist in the new set ot taints? You shouldn't need any serialization/de-serialization step in this.

> when deleting a taint from the node we rely on the annotation to know if we are allowed to touch that taint or not

Exactly! That's why we should get oldTaints from the annotation and get the diff to newTaints and see if any of those need to be deleted.

The sources of truth are:
1. The (old) annotation for *old* annotations
2. The `taints` arg (of the method) for the new annotations

Basically we just should covert the `taints` arg into the updated annotation once. (We might actually want to do some filtering there e.g. disallow any taints with a key with the `node.kubernetes.io/` prefix, for example)

>   3.2 compare taints array vs node taints to get the diff
   2.3 add/remove diff in node taint

I don't think this works. You should not remove anything that was not listed in the oldAnnotation (i.e. marked as "managed by nfd") in the first place